File: go/day1/main.go
package main

import (
	"math"
	"slices"
)

func main() {
	contents := read("input.txt")
	part1(contents)
	part2(contents)
}

func part1(contents string) {
	leftList, rightList := parse(contents)
	total := 0
	for range leftList {
		ml := minList(leftList)
		mr := minList(rightList)
		leftList = remove(leftList, ml)
		rightList = remove(rightList, mr)
		distance := max(ml, mr) - min(ml, mr)
		total += distance
	}
	println("part 1:", total)
}

func part2(contents string) {
	leftList, rightList := parse(contents)
	total := 0
	occurences := make(map[int]int)
	for _, el := range leftList {
		occurences[el]++
	}
	for _, el := range rightList {
		total += el * occurences[el]
	}
	println("part 2:", total)
}

func remove(list []int, target int) []int {
	for i, el := range list {
		if el == target {
			return append(list[:i], list[i+1:]...)
		}
	}
	return nil
}

func minList(list []int) int {
	slices.Sort(list)
	return list[0]
}

func max(x, y int) int {
	return int(math.Max(float64(x), float64(y)))
}

func min(x, y int) int {
	return int(math.Min(float64(x), float64(y)))
}

File: rust/day1/src/main.rs
use std::{collections::HashMap, fs};

fn main() {
    let contents = fs::read_to_string("input.txt").unwrap();
    let parsed = parse(&contents);
    let mut left_list = parsed.0;
    let mut right_list = parsed.1;

    left_list.sort();
    right_list.sort();

    let mut total = 0;
    for (i, _) in left_list.iter().enumerate() {
        let left = left_list.get(i).unwrap();
        let right = right_list.get(i).unwrap();
        let distance = left.max(right) - right.min(left);
        total += distance;
    }
    println!("part 1: {}", total);
    total = 0;

    let mut occurrences: HashMap<i32, i32> = HashMap::new();
    for el in left_list {
        let count = match occurrences.get(&el) {
            Some(&n) => n,
            None => 0,
        };
        occurrences.insert(el, count + 1);
    }
    for el in right_list {
        let count = match occurrences.get(&el) {
            Some(&n) => n,
            None => 0,
        };
        total += el * count
    }
    println!("part 2: {}", total);
}

fn parse(contents: &str) -> (Vec<i32>, Vec<i32>) {
    let mut left_list: Vec<i32> = Vec::new();
    let mut right_list: Vec<i32> = Vec::new();
    for line in contents.split("\n") {
        if line == "" {
            continue;
        }
        let pair = line.split("   ").collect::<Vec<&str>>();
        let left = pair.get(0).unwrap().to_string().parse::<i32>().unwrap();
        let right = pair.get(1).unwrap().to_string().parse::<i32>().unwrap();
        left_list.push(left);
        right_list.push(right);
    }
    return (left_list, right_list);
}
